tfquaternion/tfquaternion.py

```diff
@@ -163,8 +150,77 @@ def rotate_vector_by_quaternion(q, v, q_ndims=None, v_ndims=None):
         v = tf.expand_dims(v, axis=0) + tf.zeros_like(q_xyz)
     q_xyz += tf.zeros_like(v)
     v += tf.zeros_like(q_xyz)
-    t = 2 * tf.cross(q_xyz, v)
-    return v + tf.expand_dims(w, axis=-1) * t + tf.cross(q_xyz, t)
+    t = 2 * tf.linalg.cross(q_xyz, v)
+    return v + tf.expand_dims(w, axis=-1) * t + tf.linalg.cross(q_xyz, t)
+
+
+@scope_wrapper
+def get_rotation_quaternion_from_u_to_v(u, v, epsilon=1e-6, dtype=tf.float32):
+    """
+    Return a quaternion that will rotate one vector u onto another, v.
+
+    Given v amd u, this function computes q such that v = rotate_vector_by_quaternion(q, u).
+
+    Pseudocode was adapted from https://stackoverflow.com/questions/1171849/finding-
+    quaternion-representing-the-rotation-from-one-vector-to-another#:~:text=One%20
+    solution%20is%20to%20compute,all%20the%20way%20to%20v!
+
+    Args:
+    u: A `tf.Tensor` with shape (..., 3)
+    v: A `tf.Tensor` with shape (..., 3)
+    epsilon: a float, optional.  A small number used to avoid divide by zero, which
+        occurs when the rotation is a full flip (v -> -1)
+    dtype: The type used for the quaternion, must be a floating point
+        number, i.e. one of tf.float16, tf.float32, tf.float64.
+
+    Returns: A `Quaternion` with shape (..., 4)
+    """
+    u = tf.math.l2_normalize(u, axis=-1)
+    v = tf.math.l2_normalize(v, axis=-1)
+
+    # The dot / cross product determination produces twice the desired rotation.
+    # Half the rotation can be accomplished by averaging the rotation produced here with a zero
+    # rotation, which is accomplished by simply adding 1 to the dot product.
+    dot = tf.reduce_sum(tf.multiply(u, v), -1, keepdims=True) + 1
+    cross = tf.linalg.cross(u, v)
+
+    # If the rotation is a pure 180 flip then the dot product will be -1 and the cross
+    # will be (0, 0, 0), so that quaternion generated will be a zero quaterion, which is
+    # not correct.
+    # We need to construct the quaternion (0, x, y, z) where (x, y, z) is any vector
+    # orthogonal to u.  To get this, try u cross x-axis, and in the rare case where
+    # that is zero because u is already the x-axis, then crossing with the y axis is
+    # guarenteed to work, because there is no vector that can be orthogonal to both
+    x_axis = tf.cast(tf.broadcast_to((1, 0, 0), u.shape), dtype)
+    y_axis = tf.cast(tf.broadcast_to((0, 1, 0), u.shape), dtype)
+    ortho_x = tf.linalg.cross(u, x_axis)
+    ortho_y = tf.linalg.cross(u, y_axis)
+    ortho = tf.where(
+        tf.less(tf.reduce_sum(tf.multiply(ortho_x, ortho_x), -1, keepdims=True), epsilon),
```

Comment:
Could you factor out the condition? I.e.
is_parallel = tf.reduce_sum(tf.multiply(ortho_x, ortho_x), -1, keepdims=True) < epsilon
ortho = tf.where(is_parallel, ...)